Entities/MOSRotating.cpp

```diff
@@ -904,176 +951,110 @@ bool MOSRotating::ParticlePenetration(HitData &hd)
-            // Put variation on the lifetime, if it's not set to be endless
-            if (pGib->GetLifetime() != 0)
-                pGib->SetLifetime(pGib->GetLifetime() * (1.0F + ((*gItr).GetLifeVariation() * RandomNormalNum())));
-            // Set up its position and velocity according to the parameters of this AEmitter.
-            pGib->SetPos(m_Pos + gibROffset/*Vector(m_Pos.m_X + 5 * NormalRand(), m_Pos.m_Y + 5 * NormalRand())*/);
-            pGib->SetRotAngle(m_Rotation.GetRadAngle() + pGib->GetRotMatrix().GetRadAngle());
-            // Rotational angle
-            pGib->SetAngularVel((pGib->GetAngularVel() * 0.35F) + (pGib->GetAngularVel() * 0.65F / pGib->GetMass()) * RandomNum());
-            // Make it rotate away in the appropriate direction depending on which side of the object it is on
-            // If the object is far to the relft or right of the center, make it always rotate outwards to some degree
-            if (gibROffset.m_X > m_aSprite[0]->w / 3)
-            {
-                float offCenterRatio = gibROffset.m_X / (m_aSprite[0]->w / 2);
-                angularVel = fabs(pGib->GetAngularVel() * 0.5);
-                angularVel += fabs(pGib->GetAngularVel() * 0.5 * offCenterRatio);
-                pGib->SetAngularVel(angularVel * (gibROffset.m_X > 0 ? -1 : 1));
+    CreateGibsWhenGibbing(impactImpulse, movableObjectToIgnore);
+
+    RemoveAttachablesWhenGibbing(impactImpulse, movableObjectToIgnore);
+
+    m_GibSound.Play(m_Pos);
+
+    if (m_pScreenEffect && m_EffectOnGib && (m_EffectAlwaysShows || !g_SceneMan.ObscuredPoint(m_Pos.GetFloorIntX(), m_Pos.GetFloorIntY()))) {
+		g_PostProcessMan.RegisterPostEffect(m_Pos, m_pScreenEffect, m_ScreenEffectHash, 255, m_EffectRotAngle);
+    }
+
+    if (m_LoudnessOnGib > 0) {
+        g_MovableMan.RegisterAlarmEvent(AlarmEvent(m_Pos, m_Team, m_LoudnessOnGib));
+    }
+
+    m_ToDelete = true;
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void MOSRotating::CreateGibsWhenGibbing(const Vector &impactImpulse, MovableObject *movableObjectToIgnore) {
+    for (const Gib &gibSettingsObject : m_Gibs) {
+        if (gibSettingsObject.GetCount() == 0) {
+            continue;
+        }
+        MovableObject *gibParticleClone = dynamic_cast<MovableObject *>(gibSettingsObject.GetParticlePreset()->Clone());
+
+        float minVelocity = gibSettingsObject.GetMinVelocity();
+        float velocityRange = gibSettingsObject.GetMaxVelocity() - gibSettingsObject.GetMinVelocity();
+        if (gibSettingsObject.GetMinVelocity() == 0 && gibSettingsObject.GetMaxVelocity() == 0) {
+            minVelocity = m_GibBlastStrength / gibParticleClone->GetMass();
+            velocityRange = 10.0F;
+        }
+        Vector rotatedGibOffset = RotateOffset(gibSettingsObject.GetOffset());
+        for (int i = 0; i < gibSettingsObject.GetCount(); i++) {
+            if (i > 0) {
```

Comment:
Hmmm I'm not actually sure why I have this if check here. I guess it's cause I changed the `++i` to `i++` and I thought that the former would mean it started from 1 instead of 0 or something, but that's definitely not the case, so I'm just gonna remove this if check.